File: 백준/Gold/1300. K번째 수/K번째 수.py
import sys
input = sys.stdin.readline

N = int(input())
K = int(input())

left,right = 1,K

while left <= right:
    mid = (left+right)//2
    
    cnt = 0
    for i in range (1,N+1):
      cnt += min(mid//i,N)
    if cnt >= K:
        answer = mid
        right = mid-1
    else:
        left = mid+1

print(answer)

File: 백준/Gold/1300. K번째 수/K번째 수.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        int N = Integer.parseInt(br.readLine());
        int K = Integer.parseInt(br.readLine());

        long left = 1;
        long right = K;
        long answer = 0;

        while (left <= right) {
            long mid = (left + right) / 2;

            long cnt = 0;
            for (int i = 1; i <= N; i++) {
                cnt += Math.min(mid / i, N);
            }

            if (cnt >= K) {
                answer = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }

        System.out.println(answer);
    }
}